homeassistant/components/utility_meter/sensor.py

```diff
@@ -355,53 +377,70 @@ def start(self, unit):
         self._state = 0
         self.async_write_ha_state()
 
+    @staticmethod
+    def _validate_state(state: State | None) -> Decimal | None:
+        """Parse the state as a Decimal if available. Throws DecimalException if the state is not a number."""
+        return (
+            None
+            if state is None or state.state in [STATE_UNAVAILABLE, STATE_UNKNOWN]
+            else Decimal(state.state)
+        )
+
+    def calculate_adjustment(
+        self, old_state: State | None, new_state_val: Decimal
+    ) -> Decimal | None:
+        """Calculate the adjustment based on the old and (assumed valid) new state."""
+
+        if self._sensor_delta_values:
+            return new_state_val
+
+        if (
+            not self._sensor_periodically_resetting
+            and self._last_valid_state is not None
+        ):  # Fallback to old_state if sensor is periodically resetting but last_valid_state is None
+            return new_state_val - self._last_valid_state
+
+        try:
+            if (old_state_val := self._validate_state(old_state)) is not None:
+                return new_state_val - old_state_val
+        except DecimalException as err:
+            _LOGGER.warning(
+                "Invalid state (%s > %s): %s", old_state.state, new_state_val, err  # type: ignore[union-attr]  # because Exception only raised if old_state is not None
+            )
+        return None
+
     @callback
-    def async_reading(self, event):
+    def async_reading(self, event: Event):
         """Handle the sensor state changes."""
-        old_state = event.data.get("old_state")
-        new_state = event.data.get("new_state")
+        old_state: State | None = event.data.get("old_state")
+        new_state: State | None = event.data.get("new_state")
 
-        if self._state is None and new_state.state:
+        if new_state and self._state is None:
             # First state update initializes the utility_meter sensors
-            source_state = self.hass.states.get(self._sensor_source_id)
             for sensor in self.hass.data[DATA_UTILITY][self._parent_meter][
                 DATA_TARIFF_SENSORS
             ]:
-                sensor.start(source_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT))
-
-        if (
-            new_state is None
-            or new_state.state in [STATE_UNKNOWN, STATE_UNAVAILABLE]
-            or (
-                not self._sensor_delta_values
-                and (
-                    old_state is None
-                    or old_state.state in [STATE_UNKNOWN, STATE_UNAVAILABLE]
-                )
-            )
-        ):
-            return
-
-        self._unit_of_measurement = new_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
+                sensor.start(new_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT))
```

Comment:
why are you dropping source_state for new_state ?

Reply:
I'm dropping the `source_state` for the `new_state` because the result of `hass.states.get(source_entity)` is the state we received through the Event. Furthermore, if we keep using the `source_state`, we need to implement an additional None check (because the state can be None).